expertsystem/data.py

```diff
@@ -296,10 +296,14 @@ class ParticleCollection(abc.Mapping):
     def __init__(self, particles: Optional[Iterable[Particle]] = None) -> None:
         self.__particles: Dict[str, Particle] = dict()
         if particles is not None:
-            if isinstance(particles, (list, set, tuple)):
-                self.__particles.update(
-                    {particle.name: particle for particle in particles}
+            if not isinstance(particles, (list, set, tuple)):
+                raise ValueError(
+                    f"Cannot construct a {self.__class__.__name__} "
+                    f"from a {particles.__class__.__name__}"
                 )
+            self.__particles.update(
+                {particle.name: particle for particle in particles}
```

Comment:
Improvement suggestion: Actually there is no runtime check on what `particle` is. So theoretically any kind of object can be injected here as long as it has a attribute `name`.

Reply:
26fae1f1
I don't want to slow down construction too much though, that's why I put it in the comprehension